### ISSUE TITLE
notifyPath(...) with nested path and dom-repeat does not seem to work async-ly

### ISSUE THREAD

**Issue by ihab** [Polymer/polymer#2225]
I am experimenting with a custom element that listens for model changes using `Object.observe()`. I have the following code for my element:

```
<link rel="import" href="../bower_components/polymer/polymer.html">
<dom-module id="test-view">
  <template>
    <p>Hello, my name is <span>{{model.name}}</span>!</p>
    <template is="dom-repeat" items="{{model.friends}}">
      <p>I have a friend called <span>{{item.name}}</span>!</p>
    </template>
  </template>
</dom-module>

<script>
  Polymer({
    is: 'test-view',
    properties: {
      model: {
        type: Object,
        observer: '_modelChanged',
      },
      _previousModel: {
        type: Object,
      },
      _observer: {
        type: Function,
      },
    },
    publish: {
      model: { value: '', reflect: true }
    },
    ready: function() { },
    attached: function() { },
    detached: function() { },
    _modelChanged: function(model) {
      console.log('modelChanged: model =', model);
      if (!this._observer) {
        const changeResponse = (function(change) {
          console.log('notifyPath', 'change model.' + change.name, change.object[change.name]);
          this.notifyPath("model." + change.name, change.object[change.name]);
        }).bind(this);
        this._observer = function(changes) {
          changes.forEach(changeResponse);
        };
      }
      if (this._previousModel) {
        console.log('unobserving', this._previousModel);
        Object.unobserve(this._previousModel, this._observer);
      }
      if (model) {
        console.log('observing', model);
        Object.observe(model, this._observer);
      }
      this._previousModel = model;
    },
  });
</script>
```

Note the use of `notifyPath()` in response to `Object.observe()` notifications. The driver HTML looks like this:

```
<html>
  <head>
    <script src="../bower_components/webcomponentsjs/webcomponents.js"></script>
    <link rel="import" href="../test-view/test-view.html">
  </head>
  <body>
    <test-view></test-view>
    <script>
      "use strict";

      const q = (function() {
        const items = [];
        function go() {
          if (items.length > 0) { items.pop()(); }
          window.setTimeout(go, 2000);
        }
        go();
        // The following line implements async behavior
        return function(f) { items.unshift(f); }
        // The following line implements sync behavior
        // un-comment (and comment out the above) to demonstrate
        // return function(f) { f(); }
      })();

      const model = { name: 'Scooby', };
      const testView = document.querySelector('test-view');

      q(function() {
        testView.model = model;
      });

      q(function() {
        model.name = 'Fred';
      });

      q(function() {
        model.friends = [
          { name: 'Velma', },
          { name: 'Daphne', },
        ];
      });

      let x;

      q(function() {
        x = model.friends;
        model.friends = undefined;
      });

      q(function() {
        x.push({ name: 'Shaggy', });
        model.friends = x;
      });
    </script>
  </body>
</html>
```

When I run it with each change being sent synchronously (see the definition of `q()`), I get output that looks like:

    Hello, my name is Fred!
    I have a friend called Velma!
    I have a friend called Daphne!
    I have a friend called Shaggy!

When on the other hand the changes are asynchronous -- each one in a new turn of the event loop -- I get:

    Hello, my name is Fred!
    I have a friend called Velma!
    I have a friend called Daphne!
    I have a friend called !

In other words, the object `model.friends[3]` with `name` equal to `'Shaggy'` is not rendered correctly by the `dom-repeat` in my `<template>`.

Please let me know if my example is not clear. I understand it is not a minimal test case, but it is an easy platform for exercising and debugging the problem.

**Comment by ihab** [comment]
I should add that I am using Polymer 1.0.8.

**Comment by ihab** [comment]
As a more minimalistic example, please see the below element, which is self-contained and will change its own backing data by itself:

```
<link rel="import" href="../bower_components/polymer/polymer.html">
<dom-module id="test-view">
  <template>
    <p>Hello, my name is <span>{{model.name}}</span>!</p>
    <template is="dom-repeat" items="{{model.friends}}">
      <p>I have a friend called <span>{{item.name}}</span>!</p>
    </template>
  </template>
</dom-module>

<script>
  "use strict";

  Polymer({

    is: 'test-view',

    properties: {
      model: Object,
    },

    ready: function() { },

    attached: function() {
      let self = this;

      let timeout = 0;
      function q(f) {
        timeout += 2000;
        window.setTimeout(f, timeout);
      }

      q(function() {
        self.model = { name: 'Scooby', };
      });

      q(function() {
        self.model.name = 'Velma';
        self.notifyPath('model.name');
      });

      q(function() {
        self.model.friends = [ { name: 'Shaggy', } ];
        self.notifyPath('model.friends');
      });

      // The above notifyPath()-s do not work. The below
      // is required to force a model update. Comment out
      // the below to see the problem case, where the UI
      // does not refresh as I would expect it to.
      q(function() {
        const old = self.model;
        self.model = undefined;
        self.model = old;
      });
    },

    detached: function() { },
  });
</script>
```

I expect that, first, the displayed `name` should change to `'Velma'`, then there should be a `friends[0]` called `'Shaggy'` displayed. Instead, after the part that updates the name to `'Velma'`, the `{{model.name}}` portion of the `<template>` goes blank, and nothing is updated until the final function that reassigns `model` to "force" update.

Please see the code above for details -- as noted, you can comment or un-comment the last of the `q(...)` function invocations.

**Comment by kevinpschaaf** [comment]
Re: your last simplified example, that can be made to work by ensuring `notifyPath` is called with the 2nd required argument (value at that path): http://jsbin.com/sivute/edit?html,output

However, you were calling `notifyPath` in your first example correctly, so the problem there is different.  Basically, you implemented the `Object.observe` portion to correctly notify object path changes, but Polymer also requires notifying it of Array mutations (splices) to keep its internal array-item tracking in sync.  Normally this is done via the array mutation methods on polymer elements.  There is a helper API (`_notifySplices`) which is analogous to the `notifyPath` API for this purpose (although we haven't made it public), that you could theoretically call from an `Array.observe` callback that provides the proper splice information.

However, I'll point you to an experimental layer shim that we built (as a Polymer behavior) that provides a similar "action-at-a-distance" facility that you were implementing above.  It uses the [observe.js](https://github.com/polymer/observe-js) API so that it works cross-platform.
https://github.com/PolymerLabs/polymer-experiments/blob/master/observe-js-behavior.html

And for completeness, I made your first sample work using a brute-force method: copy the array (via `slice`, which will avoid Polymer using the out-of-sync array item tracking info):
http://jsbin.com/budeki/1/edit?html,output

**Comment by tomalec** [comment]
@ihab, maybe you will find [<dom-bind-notifier>](https://github.com/Juicy/dom-bind-notifier) useful, then you can write it without a worry on `notifyPath`:
```html
<template is="dom-bind" id="test-view">
    <dom-bind-notifier ref="test-view" observed-object="{{model}}" path="model" deep></dom-bind-notifier>
      <p>Hello, my name is <span>{{model.name}}</span>!</p>
      <template is="dom-repeat" items="{{model.friends}}">
        <p>I have a friend called <span>{{item.name}}</span>!</p>
      </template>
</template>
<script>
     document.querySelector('#test-view').model = model;
</script>
```

http://jsbin.com/doxade/edit?html,console,output

**Comment by ihab** [comment]
Thanks @kevinpschaaf and @tomalec -- super helpful!
